sql/expression/function/regexp_replace.go

```diff
@@ -0,0 +1,257 @@
+func (r *RegexpReplace) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	span, ctx := ctx.Span("function.RegexpReplace")
+	defer span.Finish()
+
+	cached := r.cachedVal.Load()
+	if cached != nil {
+		return cached, nil
+	}
+
+	// Evaluate string value
+	str, err := r.args[0].Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if str == nil {
+		return nil, nil
+	}
+	str, err = sql.LongText.Convert(str)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert to string
+	_str := str.(string)
+
+	// Create regex, should handle null pattern and null flags
+	r.compile(ctx, row)
+	if r.compileErr != nil {
+		return "compile bad", r.compileErr
+	}
+	if r.re == nil {
+		return nil, nil
+	}
+
+	// Evaluate ReplaceStr
+	replaceStr, err := r.args[2].Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if replaceStr == nil {
+		return nil, nil
+	}
+	replaceStr, err = sql.LongText.Convert(replaceStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert to string
+	_replaceStr := replaceStr.(string)
+
+	// Do nothing if str is empty
+	if len(_str) == 0 {
+		return _str, nil
+	}
+
+	// Check if position argument was provided
+	var pos interface{} = nil
+	if len(r.args) >= 4 {
+		pos, err = r.args[3].Eval(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+		if pos == nil {
+			return nil, nil
+		}
+	} else {
+		// Default position is 1, if argument not provided
+		pos = 1
+	}
+
+	// Convert to int32
+	pos, err = sql.Int32.Convert(pos)
+	if err != nil {
+		return nil, err
+	}
+	// Convert to int
+	_pos := int(pos.(int32))
+
+	// Non-positive position throws incorrect parameter
+	if _pos <= 0 {
+		return nil, ErrInvalidArgument.New(r.FunctionName(), fmt.Sprintf("%d", _pos))
+	}
+
+	// Handle out of bounds
+	if _pos > len(_str) {
+		return nil, errors.NewKind("Index out of bounds for regular expression search.").New()
+	}
+
+	// Check if Occurrence argument was provided
+	var occ interface{} = nil
+	if len(r.args) >= 5 {
+		occ, err = r.args[4].Eval(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+		if occ == nil {
+			return nil, nil
+		}
+	} else {
+		// Default occurrence is 0 (replace all occurrences)
+		occ = 0
+	}
+
+	// Convert occurrence to int32
+	occ, err = sql.Int32.Convert(occ)
+	if err != nil {
+		return nil, err
+	}
+	_occ := int(occ.(int32))
+
+	// MySQL interprets negative occurrences as first for some reason
+	if _occ < 0 {
+		_occ = 1
+	} else if _occ == 0 {
+		// Replace everything
+		return _str[:_pos-1] + r.re.ReplaceAllString(_str[_pos-1:], _replaceStr), nil
+	}
+
+	// Split string into prefix and suffix
+	prefix := _str[:_pos-1]
+	suffix := _str[_pos-1:]
+
+	// Extract all matches
+	matches := r.re.FindAllString(suffix, -1)
+	indexes := r.re.FindAllStringIndex(suffix, -1)
+
+	// No matches, return original string
+	if len(matches) == 0 {
+		return _str, nil
+	}
+
+	// If there aren't enough occurrences
+	if _occ > len(matches) {
+		return _str, nil
+	}
+
+	// Replace only the nth occurrence
```

Comment:
This logic is a little inscrutable but you did a good job with comments and I can't think of a better way to handle it offhand, so good work